Flaky test ElectionTest.singleLeaderConfigMapLockUpdateTest

```
Error:  Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.189 s <<< FAILURE! -- in io.fabric8.kubernetes.client.mock.LeaderElectionTest
May 07, 2026 3:35:02 PM io.fabric8.mockwebserver.MockWebServer info
Error:  io.fabric8.kubernetes.client.mock.LeaderElectionTest.singleLeaderConfigMapLockUpdateTest -- Time elapsed: 0.115 s <<< FAILURE!
INFO: MockWebServer[34791] received request: GET /api/v1/namespaces/namespace/configmaps/name HTTP/1.1 and responded: HTTP/1.1 200 OK
org.opentest4j.AssertionFailedError: expected: <lead-config-map> but was: <not-lead-config-map>
May 07, 2026 3:35:02 PM io.fabric8.mockwebserver.MockWebServer info
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
INFO: MockWebServer[34791] received request: PATCH /api/v1/namespaces/namespace/configmaps/name HTTP/1.1 and responded: HTTP/1.1 404 Not Found
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[-1799831668-pool-221-thread-1] WARN io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector - Exception occurred while acquiring lock 'ConfigMapLock: namespace - name (lead-config-map) retrying...'
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: http://localhost:34791/api/v1/namespaces/namespace/configmaps/name. Message: Not Found.
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:205)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1145)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.mock.LeaderElectionTest.testAndAssertSingleLeader(LeaderElectionTest.java:215)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:419)
	at io.fabric8.kubernetes.client.mock.LeaderElectionTest.singleLeaderConfigMapLockUpdateTest(LeaderElectionTest.java:90)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:397)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handlePatch(BaseOperation.java:776)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:231)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:236)

	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:251)
[INFO] Running io.fabric8.kubernetes.client.mock.V1CronJobTest
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:44)
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.359 s -- in io.fabric8.kubernetes.client.mock.NodeTest
	at io.fabric8.kubernetes.client.extension.ResourceAdapter.patch(ResourceAdapter.java:306)
[INFO] Running io.fabric8.kubernetes.client.mock.VisitResourcesTest
	at io.fabric8.kubernetes.client.extended.leaderelection.resourcelock.ResourceLock.update(ResourceLock.java:64)
	at io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector.tryAcquireOrRenew(LeaderElector.java:249)
	at io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector.lambda$renewWithTimeout$6(LeaderElector.java:207)
	at io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector.lambda$loop$8(LeaderElector.java:294)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: http://localhost:34791/api/v1/namespaces/namespace/configmaps/name. Message: Not Found.
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:642)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:622)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:582)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:141)
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.128 s -- in io.fabric8.kubernetes.client.mock.VisitResourcesTest
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:51)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
[INFO] Running io.fabric8.kubernetes.client.mock.WatchTest
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.vertx.VertxHttpRequest.lambda$consumeBytes$1(VertxHttpRequest.java:90)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:304)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:448)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:279)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:157)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:758)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:327)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:307)
	at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:989)
	at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:861)
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:823)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:160)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
[INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.552 s -- in io.fabric8.kubernetes.client.mock.EndpointsTest
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
[INFO] Running io.fabric8.kubernetes.client.mock.CSIDriverTest
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:361)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:325)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:171)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```

**Note: the logs are from the CI where suites are running in parallel and might include chunks from other tests**

--- a/AGENTS.md
+++ b/AGENTS.md
@@ -80,6 +80,34 @@ mvn -Pitests -pl kubernetes-itests verify
 mvn -Pitests -pl kubernetes-itests verify -Dtest="io.fabric8.openshift.**"
 ```
 
+### Reproducing CI flakes locally
+
+CI runs on 2-core GitHub Actions VMs with limited RAM. Surefire `forkCount=1C` and Vert.x event-loop sizing both scale with `Runtime.availableProcessors()`, so a flake that surfaces under CI scheduling pressure often won't reproduce with a plain `mvn test` on a 16+ core developer host. Run the test inside a constrained `docker` (or `podman`) container with the project and `~/.m2` bind-mounted — this is the closest local approximation to CI shape and almost always the fastest way to confirm or rule out a race.
+
+```bash
+# Constrained container — 2 cores, ~7 GB RAM, cgroup-enforced
+docker run --rm \
+  --cpus=2 --memory=7g --cpuset-cpus="0,1" \
+  --user $(id -u):$(id -g) -e HOME=/tmp \
+  -v "$PWD:/work" \
+  -v "$HOME/.m2:/host-m2" \
+  -w /work \
+  maven:3.9-eclipse-temurin-11 \
+  mvn -pl kubernetes-tests \
+      -Dtest='SomeFlakyTest#someFlakyMethod' \
+      -Dmaven.repo.local=/host-m2/repository \
+      test
+```
+
+For harder contention, run a sibling `stress-ng` container pinned to the same `cpuset` as a noisy neighbor, then loop the test 25–50 times under that stress.
+
+```bash
+docker run --rm -d --name stressor --cpus=2 --cpuset-cpus="0,1" \
+  alpine:latest sh -c 'apk add --no-cache stress-ng && stress-ng --cpu 4 --cpu-load 80 --timeout 600s'
+```
+
+Caveat: not every CI flake reproduces locally — GH Actions VMs add IO/network jitter that cgroup quota cannot simulate. After ~20 unsuccessful iterations, switch to instrument-and-wait (add timing logs on a debug branch and watch the next CI hit) rather than burning more local hours.
+
 ### Running Examples
 
 ```bash

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/LeaderElectionTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/LeaderElectionTest.java
@@ -192,6 +192,7 @@ class LeaderElectionTest {
     // Given
     final CountDownLatch leaderLatch = new CountDownLatch(1);
     final AtomicReference<String> newLeaderRecord = new AtomicReference<>();
+    final AtomicReference<String> leaderOnStart = new AtomicReference<>();
     final CountDownLatch stoppedLeading = new CountDownLatch(1);
     final ExecutorService executorService = Executors.newSingleThreadExecutor();
     // When
@@ -204,7 +205,10 @@ class LeaderElectionTest {
                 .withRenewDeadline(Duration.ofSeconds(10L))
                 .withRetryPeriod(Duration.ofSeconds(2L))
                 .withLeaderCallbacks(new LeaderCallbacks(
-                    leaderLatch::countDown,
+                    () -> {
+                      leaderOnStart.set(newLeaderRecord.get());
+                      leaderLatch.countDown();
+                    },
                     stoppedLeading::countDown,
                     newLeaderRecord::set))
                 .build())
@@ -212,7 +216,7 @@ class LeaderElectionTest {
             .run()));
     // Then
     assertTrue(leaderLatch.await(10, TimeUnit.SECONDS));
-    assertEquals(id, newLeaderRecord.get());
+    assertEquals(id, leaderOnStart.get());
     assertEquals(0, leaderLatch.getCount());
     leaderElectorTask.cancel(true);
     executorService.shutdownNow();

